mlflow/utils/proto_json_utils.py

```diff
@@ -417,6 +462,13 @@ def cast_schema_type(input_data):
 
     # Read the JSON
     try:
+        # objects & arrays schema for List[Dict] and Dict[List] are different
+        # so the conversion for `instances` below changes the schema.
+        # e.g.
+        # [{"col1": 1, "col2": 2}, {"col1": 3, "col2": 4}] -> {"col1": [1, 3], "col2": [2, 4]}
+        # Schema([ColSpec(long, "col1"), ColSpec(long, "col2")]) ->
+        # Schema([ColSpec(Array(long), "col1"), ColSpec(Array(long), "col2")])
+        # To avoid this, we shouldn't use `instances` for such data.
```

Comment:
Can't support `instances` here.

Reply:
For new behavior, let's raise on this breaking change with a clear error message to suggest a different input type. 
Ensure that the existing behavior does not change.

Reply:
Unfortunately we can't tell whether users are using the new schema. Now it can successfully convert the data from List[Dict] to Dict[str, List], and it sometimes fails during the following schema enforcement time. Added test here: https://github.com/mlflow/mlflow/blob/ba003f8b891a0374c7e93cfd9867307d18fa8b58/tests/pyfunc/test_pyfunc_schema_enforcement.py#L2229